src/propulsion/controller.hpp

```diff
@@ -0,0 +1,243 @@
+   */
+  void updateActualTorque();
+  /**
+   * @return int32_t - actual velocity of the motor
+   */
+
+  int32_t getVelocity() override;
+  /**
+   * @return int16_t - actual torque of the motor
+   */
+  int16_t getTorque();
+  /**
+   * @brief Set the controller to quickstop mode.
+   */
+  void quickStop() override;
+  /**
+   * @brief Check the error/warning regesters of the controller.
+   */
+  void healthCheck() override;
+  /**
+   * @return critical_failure_
+   */
+  bool getFailure() override;
+  /**
+   * @return state_
+   */
+  ControllerState getControllerState() override;
+  /**
+   * @return node_id_
+   */
+  uint8_t getNode_id();
+  /**
+   * @brief set critical_failure_
+   */
+  void setFailure(bool failure);
+  /**
+   * @brief Request the motor temperature from the controller
+   */
+  void updateMotorTemp();
+  /**
+   * @brief Request the controller temperature from the controller
+   */
+  void updateControllerTemp();
+  /**
+   * @return uint8_t - actual temperature of the motor
+   */
+  uint8_t getMotorTemp();
+  /**
+   * @return uint8_t - actual temperature of the controller
+   */
+  uint8_t getControllerTemp();
+  /**
+   * @brief to be called by processNewData if Emergency message is detected.
+   * @param message CAN message to process
+   */
+  void processEmergencyMessage(utils::io::can::Frame& message);
+  /**
+   * @brief Parses error message to find the problem
+   * @param error_message
+   */
+  void processErrorMessage(uint16_t error_message);
+  /**
+   * @brief Called by processNewData if SDO message is detected
+   * @param message
+   */
+  void processSdoMessage(utils::io::can::Frame& message);
+  /**
+   * @brief Called by processNewData if NMT message is detected
+   * @param message
+   */
+  void processNmtMessage(utils::io::can::Frame& message);
+  /*
+   * @brief { Sends state transition message to controller, leaving sufficient time for
+   *          controller to change state. If state does not change, throw critical failure }
+   *
+   * @param[in] { CAN message to be sent, Controller state requested}
+   */
+  void requestStateTransition(utils::io::can::Frame& message, ControllerState state);
+
+ private:
+  /**
+   * @brief compact function to call the can sender class with a message,
+   *        while checking for critical failure.
+   * @param message_template
+   * @param len
+   */
+  bool sendControllerMessage(ControllerMessage message_template);
+  /*
+   * @brief Sends a CAN frame but waits for a reply
+   */
+  void sendSdoMessage(utils::io::can::Frame& message);
+  /**
+   * @brief set critical failure flag to true and write failure to data structure.
+   */
+  void throwCriticalFailure();
+
+  Logger&           log_;
+  data::Data&       data_;
+  data::Motors      motor_data_;
+  ControllerState   state_;
+  uint8_t           node_id_;
+  bool              critical_failure_;
+  bool              sdo_frame_recieved_;
+  int32_t           actual_velocity_;
```

Comment:
Could you make the attributes that are changed by the processNewXYZ functions to atomic?

You only have to include <atomic> and then just write atomic<int32_t> actual_velocity and that's it